File: src/services/compliance/parser.ts
import { prisma } from "@/lib/db";
import { generateResponse } from "@/lib/ai";
import { detectContractType, type ContractTypeKey } from "./contract-types";
import {
  CLAUSE_EXTRACTION_SYSTEM_PROMPT,
  buildExtractionPrompt,
} from "./prompts";
import { CONTRACT_TYPES } from "./contract-types";
import type {
  ContractClauseKind,
  DeadlineType,
  ContractClauseMethod,
  ContractClause,
} from "@prisma/client";

// ── Types ─────────────────────────────────────────────────────────────────

interface ParsedClause {
  sectionRef: string | null;
  title: string;
  content: string;
  kind: string;
  deadlineDays: number | null;
  deadlineType: string | null;
  noticeMethod: string | null;
  trigger: string | null;
  curePeriodDays: number | null;
  curePeriodType: string | null;
  flowDownProvisions: string | null;
  parentClauseRef: string | null;
  requiresReview: boolean;
  reviewReason: string | null;
  notes: string | null;
}

interface Ambiguity {
  sectionRef: string;
  issue: string;
  recommendation: string;
}

interface ExtractionResult {
  clauses: ParsedClause[];
  ambiguities: Ambiguity[];
  contractTypeSummary?: string;
}

interface ParseContractParams {
  projectId: string;
  documentId: string;
  contractText: string;
  contractType?: ContractTypeKey;
  userId?: string;
}

interface ParseContractResult {
  clauses: ContractClause[];
  ambiguities: Ambiguity[];
  requiresReviewCount: number;
  contractType: ContractTypeKey;
  contractTypeName: string;
  tokensUsed: { input: number; output: number };
}

// ── Validation helpers ────────────────────────────────────────────────────

const VALID_KINDS: Set<string> = new Set([
  "PAYMENT_TERMS",
  "CHANGE_ORDER_PROCESS",
  "CLAIMS_PROCEDURE",
  "DISPUTE_RESOLUTION",
  "NOTICE_REQUIREMENTS",
  "RETENTION",
  "WARRANTY",
  "INSURANCE",
  "INDEMNIFICATION",
  "TERMINATION",
  "FORCE_MAJEURE",
  "LIQUIDATED_DAMAGES",
  "SCHEDULE",
  "SAFETY",
  "GENERAL_CONDITIONS",
  "SUPPLEMENTARY_CONDITIONS",
]);

const VALID_DEADLINE_TYPES: Set<string> = new Set([
  "CALENDAR_DAYS",
  "BUSINESS_DAYS",
  "HOURS",
]);

const VALID_NOTICE_METHODS: Set<string> = new Set([
  "WRITTEN_NOTICE",
  "CERTIFIED_MAIL",
  "EMAIL",
  "HAND_DELIVERY",
  "REGISTERED_MAIL",
]);

function validateKind(kind: string): ContractClauseKind {
  if (VALID_KINDS.has(kind)) return kind as ContractClauseKind;
  return "NOTICE_REQUIREMENTS";
}

function validateDeadlineType(dt: string | null): DeadlineType | null {
  if (dt && VALID_DEADLINE_TYPES.has(dt)) return dt as DeadlineType;
  return null;
}

function validateNoticeMethod(nm: string | null): ContractClauseMethod | null {
  if (nm && VALID_NOTICE_METHODS.has(nm)) return nm as ContractClauseMethod;
  return null;
}

// ── Parser ────────────────────────────────────────────────────────────────

const MAX_CONTRACT_LENGTH = 100_000; // ~25k tokens

export async function parseContract(
  params: ParseContractParams
): Promise<ParseContractResult> {
  const { projectId, documentId, contractText, userId } = params;

  // Detect contract type
  const detectedType = params.contractType ?? detectContractType(contractText);
  const contractTypeName = CONTRACT_TYPES[detectedType].name;

  // Truncate if too long
  const text = contractText.slice(0, MAX_CONTRACT_LENGTH);

  // Call Claude via shared AI service
  const aiResponse = await generateResponse({
    systemPrompt: CLAUSE_EXTRACTION_SYSTEM_PROMPT,
    userPrompt: buildExtractionPrompt(text, contractTypeName),
    model: "sonnet",
    maxTokens: 4096,
    temperature: 0.1,
  });

  // Parse JSON response — strip markdown fences if present
  let jsonStr = aiResponse.content.trim();
  if (jsonStr.startsWith("```")) {
    jsonStr = jsonStr.replace(/^```(?:json)?\n?/, "").replace(/\n?```$/, "");
  }

  let result: ExtractionResult;
  try {
    result = JSON.parse(jsonStr);
  } catch {
    throw new Error(
      `Failed to parse AI response as JSON: ${jsonStr.slice(0, 200)}`
    );
  }

  if (!Array.isArray(result.clauses)) {
    throw new Error("AI response missing clauses array");
  }

  // Persist clauses to DB
  const createdClauses: ContractClause[] = [];

  for (const parsed of result.clauses) {
    const clause = await prisma.contractClause.create({
      data: {
        projectId,
        kind: validateKind(parsed.kind),
        title: parsed.title,
        content: parsed.content || parsed.notes || "",
        sectionRef: parsed.sectionRef || null,
        deadlineDays: parsed.deadlineDays,
        deadlineType: validateDeadlineType(parsed.deadlineType),
        noticeMethod: validateNoticeMethod(parsed.noticeMethod),
        trigger: parsed.trigger || null,
        curePeriodDays: parsed.curePeriodDays,
        curePeriodType: validateDeadlineType(parsed.curePeriodType),
        flowDownProvisions: parsed.flowDownProvisions || null,
        parentClauseRef: parsed.parentClauseRef || null,
        requiresReview: parsed.requiresReview ?? false,
        reviewReason: parsed.reviewReason || null,
        confirmed: !(parsed.requiresReview ?? false),
        confirmedAt: !(parsed.requiresReview ?? false) ? new Date() : null,
        confirmedBy: !(parsed.requiresReview ?? false) ? "AI" : null,
        aiExtracted: true,
        aiModel: aiResponse.model,
        sourceDocId: documentId,
      },
    });
    createdClauses.push(clause);
  }

  // Create audit log entry
  await prisma.complianceAuditLog.create({
    data: {
      projectId,
      eventType: "CONTRACT_PARSED",
      entityType: "Document",
      entityId: documentId,
      actorType: userId ? "USER" : "SYSTEM",
      userId: userId || null,
      action: "parsed",
      details: JSON.parse(
        JSON.stringify({
          contractType: detectedType,
          contractTypeName,
          clausesFound: createdClauses.length,
          requiresReviewCount: createdClauses.filter((c) => c.requiresReview)
            .length,
          ambiguousCount: result.ambiguities?.length ?? 0,
          ambiguities: result.ambiguities ?? [],
          tokensUsed: aiResponse.tokensUsed,
          latencyMs: aiResponse.latencyMs,
        })
      ),
    },
  });

  return {
    clauses: createdClauses,
    ambiguities: result.ambiguities ?? [],
    requiresReviewCount: createdClauses.filter((c) => c.requiresReview).length,
    contractType: detectedType,
    contractTypeName,
    tokensUsed: aiResponse.tokensUsed,
  };
}

// ── Confirm / Update Clause ───────────────────────────────────────────────

interface ConfirmClauseUpdates {
  deadlineDays?: number;
  deadlineType?: string;
  noticeMethod?: string;
  trigger?: string;
  curePeriodDays?: number;
  curePeriodType?: string;
}

export async function confirmClause(
  clauseId: string,
  userId: string,
  updates?: ConfirmClauseUpdates
) {
  const clause = await prisma.contractClause.update({
    where: { id: clauseId },
    data: {
      ...(updates?.deadlineDays !== undefined && {
        deadlineDays: updates.deadlineDays,
      }),
      ...(updates?.deadlineType && {
        deadlineType: validateDeadlineType(updates.deadlineType),
      }),
      ...(updates?.noticeMethod && {
        noticeMethod: validateNoticeMethod(updates.noticeMethod),
      }),
      ...(updates?.trigger && { trigger: updates.trigger }),
      ...(updates?.curePeriodDays !== undefined && {
        curePeriodDays: updates.curePeriodDays,
      }),
      ...(updates?.curePeriodType && {
        curePeriodType: validateDeadlineType(updates.curePeriodType),
      }),
      requiresReview: false,
      confirmed: true,
      confirmedAt: new Date(),
      confirmedBy: userId,
    },
  });

  await prisma.complianceAuditLog.create({
    data: {
      projectId: clause.projectId,
      eventType: "CLAUSE_CONFIRMED",
      entityType: "ContractClause",
      entityId: clauseId,
      actorType: "USER",
      userId,
      action: "confirmed",
      details: updates ? JSON.parse(JSON.stringify({ updates })) : null,
    },
  });

  return clause;
}

// ── Query Clauses ─────────────────────────────────────────────────────────

interface ClauseFilters {
  kind?: string;
  requiresReview?: boolean;
  sourceDocId?: string;
}

export async function getProjectClauses(
  projectId: string,
  filters?: ClauseFilters
) {
  return prisma.contractClause.findMany({
    where: {
      projectId,
      ...(filters?.kind && { kind: filters.kind as ContractClauseKind }),
      ...(filters?.requiresReview !== undefined && {
        requiresReview: filters.requiresReview,
      }),
      ...(filters?.sourceDocId && { sourceDocId: filters.sourceDocId }),
    },
    orderBy: { createdAt: "desc" },
  });
}

File: backend/app/services/compliance/parser.py
"""Contract clause extraction via Claude AI.

Parses contract documents to extract compliance-critical clauses using Claude.
Stores extracted clauses in ContractClause model and creates an audit log entry.
"""

import json
import logging

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.compliance import ComplianceAuditLog, ContractClause
from app.models.document import Document
from app.models.enums import ContractClauseKind, ContractClauseMethod, DeadlineType
from app.services.ai import generate_response

from .prompts import CONTRACT_EXTRACTION_SYSTEM, CONTRACT_EXTRACTION_USER

logger = logging.getLogger(__name__)

# Valid enum values for validation
VALID_KINDS = {e.value for e in ContractClauseKind}
VALID_METHODS = {e.value for e in ContractClauseMethod}
VALID_DEADLINE_TYPES = {e.value for e in DeadlineType}


async def extract_clauses_from_document(
    db: AsyncSession,
    project_id: str,
    document_id: str,
    user_id: str | None = None,
) -> list[ContractClause]:
    """Extract compliance clauses from a document using Claude AI.

    Args:
        db: Database session.
        project_id: Project ID.
        document_id: Document to parse.
        user_id: User who initiated the extraction.

    Returns:
        List of created ContractClause records.
    """
    # Load document
    result = await db.execute(
        select(Document).where(
            Document.id == document_id,
            Document.project_id == project_id,
        )
    )
    document = result.scalar_one_or_none()
    if not document:
        raise ValueError(f"Document {document_id} not found in project {project_id}")

    # Get document text from chunks
    from app.models.document import DocumentChunk

    chunk_result = await db.execute(
        select(DocumentChunk.content)
        .where(DocumentChunk.document_id == document_id)
        .order_by(DocumentChunk.chunk_index)
    )
    chunks = chunk_result.scalars().all()
    if not chunks:
        raise ValueError(f"Document {document_id} has no text chunks")

    document_text = "\n\n".join(chunks)

    # Truncate if too long (Claude token budget)
    max_chars = 100_000
    if len(document_text) > max_chars:
        document_text = document_text[:max_chars] + "\n\n[... truncated ...]"

    # Build prompt
    user_prompt = CONTRACT_EXTRACTION_USER.format(
        document_name=document.name,
        document_type=document.type.value if document.type else "CONTRACT",
        document_text=document_text,
    )

    # Call Claude (use opus for complex extraction)
    ai_response = generate_response(
        system_prompt=CONTRACT_EXTRACTION_SYSTEM,
        user_prompt=user_prompt,
        model="opus",
        max_tokens=8000,
        temperature=0.1,
    )

    # Parse response
    raw_clauses = _parse_clause_json(ai_response.content)
    if not raw_clauses:
        logger.warning("No clauses extracted from document %s", document_id)
        return []

    # Delete existing AI-extracted clauses for this document (re-extraction)
    existing = await db.execute(
        select(ContractClause).where(
            ContractClause.project_id == project_id,
            ContractClause.source_doc_id == document_id,
            ContractClause.ai_extracted == True,  # noqa: E712
        )
    )
    for old_clause in existing.scalars().all():
        await db.delete(old_clause)

    # Create clause records
    created: list[ContractClause] = []
    for raw in raw_clauses:
        clause = _build_clause(raw, project_id, document_id, ai_response.model)
        if clause:
            db.add(clause)
            created.append(clause)

    # Audit log
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="CLAUSE_EXTRACTION",
        entity_type="Document",
        entity_id=document_id,
        user_id=user_id,
        actor_type="AI",
        action="extract_clauses",
        details={
            "documentName": document.name,
            "clausesExtracted": len(created),
            "model": ai_response.model,
            "tokensUsed": ai_response.tokens_used,
        },
    )
    db.add(audit)

    await db.flush()
    logger.info(
        "Extracted %d clauses from document %s (project %s)",
        len(created), document_id, project_id,
    )

    return created


async def get_clauses_for_project(
    db: AsyncSession,
    project_id: str,
    kind: ContractClauseKind | None = None,
    confirmed_only: bool = False,
) -> list[ContractClause]:
    """Get contract clauses for a project."""
    query = select(ContractClause).where(
        ContractClause.project_id == project_id,
    ).order_by(ContractClause.created_at.desc())

    if kind:
        query = query.where(ContractClause.kind == kind)
    if confirmed_only:
        query = query.where(ContractClause.confirmed == True)  # noqa: E712

    result = await db.execute(query)
    return list(result.scalars().all())


async def confirm_clause(
    db: AsyncSession,
    clause_id: str,
    project_id: str,
    user_id: str,
) -> ContractClause | None:
    """Confirm a clause (mark as reviewed and accurate)."""
    result = await db.execute(
        select(ContractClause).where(
            ContractClause.id == clause_id,
            ContractClause.project_id == project_id,
        )
    )
    clause = result.scalar_one_or_none()
    if not clause:
        return None

    clause.confirmed = True
    clause.confirmed_at = func.now()
    clause.confirmed_by = user_id
    clause.requires_review = False

    # Audit log
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="CLAUSE_CONFIRMED",
        entity_type="ContractClause",
        entity_id=clause_id,
        user_id=user_id,
        actor_type="USER",
        action="confirm_clause",
        details={"clauseTitle": clause.title, "clauseKind": clause.kind.value},
    )
    db.add(audit)

    await db.flush()
    return clause


def _parse_clause_json(content: str) -> list[dict]:
    """Parse AI response into a list of clause dicts."""
    # Strip markdown code fences if present
    text = content.strip()
    if text.startswith("```"):
        lines = text.split("\n")
        # Remove first and last lines (```json and ```)
        lines = [l for l in lines if not l.strip().startswith("```")]
        text = "\n".join(lines)

    try:
        parsed = json.loads(text)
        if isinstance(parsed, list):
            return parsed
        if isinstance(parsed, dict) and "clauses" in parsed:
            return parsed["clauses"]
        return []
    except json.JSONDecodeError:
        # Try to find JSON array in the response
        start = text.find("[")
        end = text.rfind("]")
        if start != -1 and end != -1:
            try:
                return json.loads(text[start : end + 1])
            except json.JSONDecodeError:
                pass
        logger.error("Failed to parse clause extraction response")
        return []


def _build_clause(
    raw: dict,
    project_id: str,
    document_id: str,
    model: str,
) -> ContractClause | None:
    """Build a ContractClause from raw AI output, with validation."""
    kind_str = raw.get("kind", "")
    if kind_str not in VALID_KINDS:
        logger.warning("Invalid clause kind: %s", kind_str)
        return None

    title = raw.get("title", "")
    content = raw.get("content", "")
    if not title or not content:
        return None

    # Validate optional enums
    notice_method = raw.get("noticeMethod")
    if notice_method and notice_method not in VALID_METHODS:
        notice_method = None

    deadline_type = raw.get("deadlineType")
    if deadline_type and deadline_type not in VALID_DEADLINE_TYPES:
        deadline_type = None

    cure_period_type = raw.get("curePeriodType")
    if cure_period_type and cure_period_type not in VALID_DEADLINE_TYPES:
        cure_period_type = None

    return ContractClause(
        project_id=project_id,
        kind=ContractClauseKind(kind_str),
        title=title,
        content=content,
        section_ref=raw.get("sectionRef"),
        deadline_days=_safe_int(raw.get("deadlineDays")),
        deadline_type=DeadlineType(deadline_type) if deadline_type else None,
        notice_method=ContractClauseMethod(notice_method) if notice_method else None,
        trigger=raw.get("trigger"),
        cure_period_days=_safe_int(raw.get("curePeriodDays")),
        cure_period_type=DeadlineType(cure_period_type) if cure_period_type else None,
        flow_down_provisions=raw.get("flowDownProvisions"),
        parent_clause_ref=raw.get("parentClauseRef"),
        requires_review=raw.get("requiresReview", False),
        review_reason=raw.get("reviewReason"),
        ai_extracted=True,
        ai_model=model,
        source_doc_id=document_id,
    )


def _safe_int(val) -> int | None:
    """Safely convert a value to int."""
    if val is None:
        return None
    try:
        return int(val)
    except (ValueError, TypeError):
        return None
